python/ray/util/sgd/torch/examples/gat_dgl.py

```diff
@@ -0,0 +1,274 @@
+                 activation, feat_drop, attn_drop, negative_slope, residual):
+        super().__init__()
+
+        self.n_layers = n_layers
+        self.activation = activation
+        self.n_hidden = n_hidden
+        self.n_heads = n_heads
+        self.n_classes = n_classes
+        self.convs = nn.ModuleList()
+
+        # input layer
+        self.convs.append(
+            GATConv((in_feats, in_feats), n_hidden, n_heads, feat_drop,
+                    attn_drop, negative_slope, residual, self.activation))
+        # hidden layer
+        for l in range(1, n_layers - 1):
+            # due to multi-head, the in_dim = num_hidden * num_heads
+            self.convs.append(
+                GATConv((n_hidden * n_heads, n_hidden * n_heads), n_hidden,
+                        n_heads, feat_drop, attn_drop, negative_slope,
+                        residual, self.activation))
+        ##output layer
+        self.convs.append(
+            GATConv((n_hidden * n_heads, n_hidden * n_heads), n_classes,
+                    n_heads, feat_drop, attn_drop, negative_slope, residual,
+                    None))
+
+    def forward(self, blocks, x):
+        h = x
+        for l, (layer, block) in enumerate(zip(self.convs, blocks)):
+            h_dst = h[:block.number_of_dst_nodes()]
+            if l != len(self.convs) - 1:
+                h = layer(block, (h, h_dst)).flatten(1)
+                h = F.dropout(h, p=0.5, training=self.training)
+            else:
+                h = layer(block, (h, h_dst))
+        h = h.mean(1)
+        return h.log_softmax(dim=-1)
+
+
+def compute_acc(pred, labels):
+    """
+    Compute the accuracy of prediction given the labels.
+    """
+    return (th.argmax(pred, dim=1) == labels).float().sum() / len(pred)
+
+
+class _NodeCollator(NodeCollator):
+    def collate(self, items):
+        # input_nodes, output_nodes, blocks
+        result = super().collate(items)
+        return result
+
+
+class CustomTrainingOperator(TrainingOperator):
+    def setup(self, config):
+        # load reddit data
+        data = RedditDataset()
+        g = data[0]
+        g.ndata['features'] = g.ndata['feat']
+        g.ndata['labels'] = g.ndata['label']
+        self.in_feats = g.ndata['features'].shape[1]
+        self.n_classes = data.num_classes
+        # add self loop,
+        g = dgl.remove_self_loop(g)
+        g = dgl.add_self_loop(g)
+        # Create csr/coo/csc formats before launching training processes
+        g.create_formats_()
+        self.g = g
+        train_nid = th.nonzero(g.ndata['train_mask'], as_tuple=True)[0]
+        val_nid = th.nonzero(g.ndata['val_mask'], as_tuple=True)[0]
+        test_nid = th.nonzero(g.ndata['test_mask'], as_tuple=True)[0]
+        self.train_nid = train_nid
+        self.val_nid = val_nid
+        self.test_nid = test_nid
+
+        # Create sampler
+        sampler = dgl.dataloading.MultiLayerNeighborSampler(
+            [int(fanout) for fanout in args.fan_out.split(',')])
```

Comment:
See comment below, but I would recommend not accessing `args` directly here, and instead pass everything through `config`.

Reply:
OK, All parameters in the custom class have been changed to pass by config.